_posts/2020-11-03-torch-tabular/torch_tabular.Rmd

```diff
@@ -0,0 +1,445 @@
+    toc: true
+preview: images/preview.jpeg
+---
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+Machine learning on image-like data can be many things: Fun (dogs vs. cats), societally useful (medical imaging), or societally harmful (surveillance). In comparison, tabular data -- the bread and butter of data science -- may seem more mundane.
+
+What's more, if you're interested in deep learning (DL) particularly, and looking for the extra benefits to be gotten from big data, big architectures, and big compute, you're much more likely to build an impressive showcase on the former instead of the latter.
+
+So for tabular data, why not just go with random forests, or gradient boosting, or other classical methods? I can think of at least a few reasons to learn about DL for tabular data:
+
+-   Even if all your features are interval-scale or ordinal, thus requiring "just" some form of (not necessarily linear) regression, applying DL may result in performance benefits due to sophisticated optimization algorithms, activation functions, layer depth, and more (plus interactions of all of these).
+
+-   If, in addition, there are categorical features, DL models may profit from *embedding* those in continuous space, discovering similarities and relationships that go unnoticed in one-hot encoded representations.
+
+-   What if most features are numeric or categorical, but there's also text in column F and an image in column G? With DL, different modalities can be worked on by different modules that feed their outputs into a common module, to take over from there.
+
+## Agenda
+
+In this introductory post, we keep the architecture straightforward. We don't experiment with fancy optimizers or nonlinearities. However, we do make use of embeddings, and pretty prominently at that. Thus from the above bullet list, we'll shed a light on the second, while leaving the other two for future posts.
+
+In a nutshell, what we'll see is
+
+-   How to create a custom *dataset*, tailored to the specific data you have.
+
+-   How to handle a mix of numeric and categorical data.
+
+-   How to extract the continuous-space representations from the embedding modules.
+
+## Dataset
+
+The dataset, [Mushrooms](https://archive.ics.uci.edu/ml/datasets/Mushroom), was chosen for its abundance of categorical columns. It is an unusual dataset to use in DL: It was designed for machine learning models to infer logical rules, as in: IF *a* AND NOT *b* OR *c* [...], then it's an *x*.
+
+Concretely, mushrooms are classified into two groups: edible and non-edible. The dataset description lists five possible rules with their resulting accuracies. While the least we want to go into here is the hotly debated topic of whether DL is suited to, or how it could be made more suited to, rule learning, we'll allow ourselves some curiosity and check out what happens if we successively remove all columns used to construct those five rules.
```

Comment:
```suggestion
Mushrooms are classified into two groups: edible and non-edible. The dataset description lists five possible rules with their resulting accuracies. While the least we want to go into here is the hotly debated topic of whether DL is suited to, or how it could be made more suited to rule learning, we'll allow ourselves some curiosity and check out what happens if we successively remove all columns used to construct those five rules.
```